state/state.go

```diff
@@ -38,9 +38,10 @@ type State struct {
 	mtx sync.Mutex
 	db  dbm.DB
 
-	// should not change
-	GenesisDoc *types.GenesisDoc
-	ChainID    string
+	// See https://github.com/tendermint/tendermint/issues/671.
```

Comment:
I know Jae doesn't like private mutable data fields in the State. We may need to seek another way to do this ... perhaps the params need to live in the ConsensusState.

Immutable private chainID should be ok.

Reply:
> perhaps the params need to live in the ConsensusState.

Will these params always be for **consensus only** (maybe we'll want to add blockchain params or something?).

Reply:
Also, if I am not mistaken, we do not persist ConsensusState (as we recreate it using blockstore & WAL). So, how about we store ChainID & Params under a separate key (`genesisParamsKey`)?

Reply:
Ok, I think we should just expose ChainID and Params in here and let them be written every time too. 

I'd ask @jaekwon to confirm.

Also if we will want historical params like we have historical validators once we update ABCI to allow changes there.

It would also be amazing if state encoding was extensible so we dont have a breaking change every time we want to change something here!

Reply:
Yeah we should definitely expose ChainID.
ConsensusParams is small enough, we can just write it.